### ISSUE TITLE
(Bounty) Add MIDI Out to ScummVM and px68k

### ISSUE THREAD

**Issue by anothername99** [libretro/RetroArch#6908]
This is for both cores because very little code is likely necessary. You can use this Dosbox commit as a reference: https://github.com/libretro/dosbox-libretro/commit/e6802ac189c05edcb71abe5b473771415ff65ecf

**Comment by ToniBC** [comment]
Other emulators such as Dosbox Pure, Scummvm and some others use Fluidsynth for the Midi, that is, the code for that is already there and implementing it does not have to be complicated, with that it would solve the problems of the MIDI of this core. Obviously using Soundfonts.

**Comment by LibretroAdmin** [comment]
I believe MIDI support has been implemented for PX68K core now? Is it sufficient?

**Comment by ToniBC** [comment]
If it's already implemented, I'll test it to see. It's a simple solution, but the best would be to use the same as Scummvm or Dosbox, the Fluidsync directly in the core.

**Comment by ectorhga** [comment]
Oh, wow! I just stumbled over this here. MIDI support for this core sounds amazing! I'll try that, too. Any hints on how I have to set this up?
Do I need MUNT in the background and pass the MIDI signals through ALSA ports?

**Comment by MrBlastman** [comment]
MIDI is not working for me at all.  I have an external SC-55 and MT-32 and neither register anything when starting games such as Akumajou Dracula.

**Comment by MrBlastman** [comment]
This is not working flawlessly.  In the current implementation it is not setting sound channels properly when GS devices such as the SC-55 are set to GS mode for games such as Akumajou Dracula--which require GS mode.  They sound broken without passing this data to the device.

**Comment by riggles1** [comment]
> This is not working flawlessly. In the current implementation it is not setting sound channels properly when GS devices such as 

I can't get it to work either, it's all turning into pianos. That's what's happening on your end as well?

**Comment by biffrapper** [comment]
> > This is not working flawlessly. In the current implementation it is not setting sound channels properly when GS devices such as
> 
> I can't get it to work either, it's all turning into pianos. That's what's happening on your end as well?

Yep, that is exactly what is happening. It is not properly initing instruments/GS.

**Comment by Kelvfimer** [comment]
I confirm that the Midi support is a bit glitchy in all games.

Besides if you play Arcus Odissey or Sol-Feace in midi mode (press 登 key when bootting that in the emu is mapped to scroll lock key) using MT-32 and SC-55 the games doesn't recognize SC-55. 

On the other hand, I may say that using MT-32 munt, SC55 ZZdenis sf2 and a custom SC-55.cfg in timidity I had almost similar midi music as the original X68000 for Castlevania.

So pls can anybody implement proper Midi Support :) thx!!

**Comment by Kelvfimer** [comment]
BTW, I opened a feature request in MAME libretro for MIDI support, regular MAME has MIDI support but libretro one nope. It's good to have several options. The only thing needed is somebody with knowledge to make the MIDI Wrapper as the one done in PC68k libretro.

https://github.com/libretro/mame/issues/477

**Comment by Starman99x** [comment]
MIDI is still broken in this core with games like Akumajou Dracula.

**Comment by Kelvfimer** [comment]
@Starman99x use px68k it has midi support. I implemented in emuelec

**Comment by Starman99x** [comment]
@Kelvfimer hmm I could not get LA or GS to work with Akumajou Dracula:

![Image](https://github.com/user-attachments/assets/f1855874-348c-4670-b579-76431040ad66)

I changed core options accordingly and also tried to enable MIDI in RetroArch/Audio settings.

**Comment by biffrapper** [comment]
> [@Kelvfimer](https://github.com/Kelvfimer) hmm I could not get LA or GS to work with Akumajou Dracula:
> 
> ![Image](https://github.com/user-attachments/assets/f1855874-348c-4670-b579-76431040ad66)
> 
> I changed core options accordingly and also tried to enable MIDI in RetroArch/Audio settings.


What LA or GS devices are you using on your PC?

Since Px68k was updated and fixed, both of these work flawlessly, both on external devices, and through MUNT and Nuked SC-55.

To use the latter, one must also install the program "loopMIDI," and then when starting MUNT or SC-55, selected loopMIDI as the midi output port.

**Comment by biffrapper** [comment]
![Image](https://github.com/user-attachments/assets/afcef2b2-21ad-4de5-bd28-4ff3c5796264)

Working great ^^

**Comment by Starman99x** [comment]
I'll see what is going on in my setup. Does GS work for you too? I'm having trouble initiating anything other than Microsoft GS. I'll try loopmidi.

**Comment by biffrapper** [comment]
> I'll see what is going on in my setup. Does GS work for you too? I'm having trouble initiating anything other than Microsoft GS. I'll try loopmidi.

Google "Nuked SC-55"

Get that, use with Loopmidi.  Follow the instructions.  Is pretty easy to set up.

**Comment by ma-moon** [comment]
I've implemented MIDI Out support for ScummVM libretro core.

## Implementation
- **ScummVM**: PR submitted at https://github.com/libretro/scummvm/pull/95
  - Added MidiDriver_Libretro class in backends/midi/libretro.cpp
  - Integrated MIDI interface initialization in retro_init()
  - Added MIDI flush in retro_run()
  
- **px68k**: Already has MIDI Out support implemented (libretro.c lines 294-363, 1855)

## Reference Implementation
Based on the dosbox-libretro commit: https://github.com/libretro/dosbox-libretro/commit/e6802ac189c05edcb71abe5b473771415ff65ecf

The implementation uses RETRO_ENVIRONMENT_GET_MIDI_INTERFACE to get the MIDI interface from the frontend and sends MIDI messages through it. This allows ScummVM to output MIDI to external synthesizers configured in RetroArch.

Please review the PR and let me know if any changes are needed.
